src/utils/resources.py

```diff
@@ -1608,6 +1609,142 @@ def deprovision(self, context: Dict = {}):
+        db_names = [name for name in properties]
+        alt_names = self._get_all_alt_names(properties)
+        colliding_names = set(db_names) & set(alt_names)
+        if colliding_names:
+            raise YunohostError(
+                f'Cannot define properties whose names are also specified as previous_names: {", ".join(colliding_names)}',
+                raw_msg=True
+            )
+        super().__init__({"redis": properties}, *args, **kwargs)
+
+    def _ensure_array(self, val: str|list[str]):
+        return [val] if isinstance(val, str) else val
+
+    def _get_all_alt_names(self, properties: Dict[str, Any]):
+        return [
+            alt_name
+            for subprops in properties.values()
+                for alt_name in self._ensure_array(subprops.get('previous_names', []))
+        ]
+
+    def _free_db(self, count=1):
+        info = check_output(
+            "redis-cli INFO keyspace"
+        )
+        databases_num_str = re.findall(r'^db(\d+):', info, flags=re.MULTILINE)
+        databases_num = [int(num) for num in databases_num_str]
+        redis_conf_path='/etc/redis/redis.conf'
+        max_dbs_match = searchf(r'^databases\s+(\d+)', redis_conf_path, count=1, flags=re.MULTILINE)
+        if max_dbs_match is None:
+            raise YunohostError(
+                f"Can't read the maximal number of Redis database allowed in {redis_conf_path}",
+                raw_msg=True
+            )
+        max_dbs = int(max_dbs_match)
+        free_dbs = set(range(0, max_dbs)) - set(databases_num)
+        if len(free_dbs) < count:
+            raise YunohostError(
+                f'Uhoh, you seem to have reached the maximum of Redis databases allowed. You may increase the value of "database" in {redis_conf_path} to increase it.',
+                raw_msg=True
+            )
+        return list(free_dbs)[:count]
+
+    def _deprovision_db(self, db_name):
+        os.system(f'redis-cli -n "{self.get_setting(db_name)}" flushdb >/dev/null 2>/dev/null')
+        self.delete_setting(db_name)
+
+    def provision_or_update(self, context: Dict = {}):
+        db_names = [name for name in self.redis]
+        dbs_to_provision = set(db_names)
+        for db_name in db_names:
+            # First check whether the env variable exist in the app's settings
+            if self.get_setting(db_name) is not None:
+                dbs_to_provision.remove(db_name)
+                continue
+            previous_names = self.redis.get(db_name, {}).get('previous_names', [])
+            for alt_name in self._ensure_array(previous_names):
+                db_num = self.get_setting(alt_name)
+                if db_num is not None:
+                    self.set_setting(db_name, db_num)
+                    self.delete_setting(alt_name)
+                    dbs_to_provision.remove(db_name)
+                    break
+
+        previously_provisioned_db = set(str(self.get_setting('provisionned_redis_db_names') or '').split(","))
```

Comment:
I am not very fund of `provisionned_redis_db_names`, I feel that it is a bit awkward. But it sounds like a viable hack.

I have proposed an alternate implementation in #2124 .

Please let me know what approach sounds the best to you, and I'll keep only one PR.